File: Packages/WebviewRpc/Runtime/WebViewRpcServer.cs
using System;
using System.Collections.Generic;
using System.Threading;
using Cysharp.Threading.Tasks;
using Google.Protobuf;
using UnityEngine;

namespace WebViewRPC
{
    public class WebViewRpcServer : IDisposable
    {
        private bool _disposed;
        
        private readonly IWebViewBridge _bridge;
        private readonly ChunkAssembler _chunkAssembler = new();
        private readonly CancellationTokenSource _cancellationTokenSource = new();

        /// <summary>
        /// You can add multiple services to the server.
        /// </summary>
        public List<ServiceDefinition> Services { get; } = new();
        
        private readonly Dictionary<string, Func<ByteString, UniTask<ByteString>>> _methodHandlers = new();

        public WebViewRpcServer(IWebViewBridge bridge)
        {
            _bridge = bridge;
            _bridge.OnMessageReceived += OnBridgeMessage;
        }

        /// <summary>
        /// Register ServiceDefinition list to actual method handlers
        /// </summary>
        public void Start()
        {
            foreach (var service in Services)
            {
                foreach (var handler in service.MethodHandlers)
                {
                    _methodHandlers[handler.Key] = handler.Value;
                }
            }
        }

        private async void OnBridgeMessage(string base64)
        {
            if (_disposed) return;

            try
            {
                var bytes = Convert.FromBase64String(base64);
                var envelope = RpcEnvelope.Parser.ParseFrom(bytes);

                // Check if this is a chunked message
                if (envelope.ChunkInfo != null)
                {
                    // Try to reassemble
                    var completeData = _chunkAssembler.TryAssemble(envelope);
                    if (completeData != null)
                    {
                        // Create a new envelope with the complete data
                        var completeEnvelope = new RpcEnvelope
                        {
                            RequestId = envelope.RequestId,
                            IsRequest = envelope.IsRequest,
                            Method = envelope.Method,
                            Payload = ByteString.CopyFrom(completeData)
                        };
                        
                        // Only set Error if it's not null or empty
                        if (!string.IsNullOrEmpty(envelope.Error))
                        {
                            completeEnvelope.Error = envelope.Error;
                        }
                        
                        if (completeEnvelope.IsRequest)
                        {
                            await HandleRequestAsync(completeEnvelope, _cancellationTokenSource.Token);
                        }
                    }
                    // else: waiting for more chunks
                }
                else
                {
                    // Process as regular message
                    if (envelope.IsRequest)
                    {
                        await HandleRequestAsync(envelope, _cancellationTokenSource.Token);
                    }
                }
            }
            catch (OperationCanceledException)
            {
                // Expected when disposing
                Debug.Log($"RPC operation cancelled for disposed server");
            }
            catch (Exception ex)
            {
                Debug.LogError($"Exception while handling message: {ex}");
            }
        }

        private async UniTask HandleRequestAsync(RpcEnvelope requestEnvelope, CancellationToken cancellationToken)
        {
            // Check disposed state before processing
            if (_disposed || cancellationToken.IsCancellationRequested) return;
            
            ByteString responsePayload = null;
            string error = null;

            try
            {
                if (_methodHandlers.TryGetValue(requestEnvelope.Method, out var handler))
                {
                    responsePayload = await handler(requestEnvelope.Payload);
                }
                else
                {
                    error = $"Unknown method: {requestEnvelope.Method}";
                }
            }
            catch (Exception ex)
            {
                error = ex.Message;
            }

            // Check disposed state before sending response
            if (_disposed || cancellationToken.IsCancellationRequested) return;

            // Send response
            // Check if this is an error response
            if (!string.IsNullOrEmpty(error))
            {
                // Error response
                var responseEnvelope = new RpcEnvelope
                {
                    RequestId = requestEnvelope.RequestId,
                    IsRequest = false,
                    Method = requestEnvelope.Method,
                    Error = error
                };
                
                // Include payload if available (even if empty)
                if (responsePayload != null)
                {
                    responseEnvelope.Payload = responsePayload;
                }
                
                var bytes = responseEnvelope.ToByteArray();
                var base64 = Convert.ToBase64String(bytes);
                _bridge.SendMessageToWeb(base64);
            }
            else if (responsePayload != null)
            {
                // Success response
                var responseBytes = responsePayload.ToByteArray();
                
                // Check if chunking is needed
                int effectivePayloadSize = WebViewRpcConfiguration.GetEffectivePayloadSize();
                if (WebViewRpcConfiguration.EnableChunking && 
                    responseBytes.Length > effectivePayloadSize)
                {
                    // Send as chunks
                    await SendChunkedMessage(requestEnvelope.RequestId, requestEnvelope.Method, 
                        responseBytes, false, null, cancellationToken);
                }
                else
                {
                    // Send as single message
                    var responseEnvelope = new RpcEnvelope
                    {
                        RequestId = requestEnvelope.RequestId,
                        IsRequest = false,
                        Method = requestEnvelope.Method,
                        Payload = responsePayload
                    };
                    
                    var bytes = responseEnvelope.ToByteArray();
                    var base64 = Convert.ToBase64String(bytes);
                    _bridge.SendMessageToWeb(base64);
                }
            }
            else
            {
                // No payload and no error - this is an error condition
                var responseEnvelope = new RpcEnvelope
                {
                    RequestId = requestEnvelope.RequestId,
                    IsRequest = false,
                    Method = requestEnvelope.Method,
                    Error = "Method returned null without error"
                };
                
                var bytes = responseEnvelope.ToByteArray();
                var base64 = Convert.ToBase64String(bytes);
                _bridge.SendMessageToWeb(base64);
            }
        }
        
        private async UniTask SendChunkedMessage(string requestId, string method, byte[] data, 
            bool isRequest, string error = null, CancellationToken cancellationToken = default)
        {
            // Check disposed state before starting
            if (_disposed || cancellationToken.IsCancellationRequested) return;
            
            var chunkSetId = $"{requestId}_{Guid.NewGuid():N}";
            int effectivePayloadSize = WebViewRpcConfiguration.GetEffectivePayloadSize();
            var totalChunks = (int)Math.Ceiling((double)data.Length / effectivePayloadSize);
            
            for (int i = 1; i <= totalChunks; i++)
            {
                // Check disposed state before each chunk
                if (_disposed || cancellationToken.IsCancellationRequested) return;
                
                var offset = (i - 1) * effectivePayloadSize;
                var length = Math.Min(effectivePayloadSize, data.Length - offset);
                var chunkData = new byte[length];
                Array.Copy(data, offset, chunkData, 0, length);
                
                var envelope = new RpcEnvelope
                {
                    RequestId = requestId,
                    IsRequest = isRequest,
                    Method = method,
                    Payload = ByteString.CopyFrom(chunkData),
                    ChunkInfo = new ChunkInfo
                    {
                        ChunkSetId = chunkSetId,
                        ChunkIndex = i,
                        TotalChunks = totalChunks,
                        OriginalSize = data.Length
                    }
                };
                
                // Only set error on the first chunk
                if (i == 1 && !string.IsNullOrEmpty(error))
                {
                    envelope.Error = error;
                }
                
                var bytes = envelope.ToByteArray();
                var base64 = Convert.ToBase64String(bytes);
                _bridge.SendMessageToWeb(base64);
            }
            
            await UniTask.CompletedTask;
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                _disposed = true;
                
                // Cancel all pending operations
                _cancellationTokenSource?.Cancel();
                _cancellationTokenSource?.Dispose();
                
                _bridge.OnMessageReceived -= OnBridgeMessage;
                
                // Note: Do not dispose the bridge here as it may be shared
                // The owner of the bridge should dispose it
            }
        }
    }
}

File: webview_rpc_runtime_library/src/core/webview_rpc_server.js
import { decodeRpcEnvelope, encodeRpcEnvelope } from './rpc_envelope.js';
import { base64ToUint8Array, uint8ArrayToBase64 } from './webview_rpc_utils.js';
import { WebViewRpcConfiguration } from './webview_rpc_configuration.js';
import { ChunkAssembler } from './chunk_assembler.js';

/**
 * WebView RPC Server
 * Handles incoming RPC requests from Unity
 */
export class WebViewRpcServer {
    constructor(bridge) {
        this._bridge = bridge;
        this._services = [];
        this._methodHandlers = {};
        this._chunkAssembler = new ChunkAssembler();
        
        // Listen for messages from Unity
        this._bridge.onMessage((base64Message) => {
            this._handleMessage(base64Message);
        });
    }

    /**
     * Add a service to the server
     * @param {ServiceDefinition} service 
     */
    addService(service) {
        this._services.push(service);
    }

    /**
     * Start the server and register all method handlers
     */
    start() {
        // Merge all service handlers into one map
        for (const service of this._services) {
            for (const [methodName, handler] of Object.entries(service.methodHandlers)) {
                this._methodHandlers[methodName] = handler;
            }
        }
    }

    /**
     * Handle incoming message from Unity
     * @param {string} base64Message 
     */
    async _handleMessage(base64Message) {
        try {
            const bytes = base64ToUint8Array(base64Message);
            const envelope = decodeRpcEnvelope(bytes);

            // Check if this is a chunked message
            if (envelope.chunkInfo) {
                // Try to reassemble
                const completeData = this._chunkAssembler.tryAssemble(envelope);
                if (completeData) {
                    // Create a new envelope with the complete data
                    const completeEnvelope = {
                        requestId: envelope.requestId,
                        isRequest: envelope.isRequest,
                        method: envelope.method,
                        payload: completeData,
                        error: envelope.error
                    };
                    
                    if (completeEnvelope.isRequest) {
                        await this._handleRequest(completeEnvelope);
                    }
                }
                // else: waiting for more chunks
            } else {
                // Process as regular message
                if (envelope.isRequest) {
                    await this._handleRequest(envelope);
                }
            }
        } catch (error) {
            console.error('Error handling message:', error);
        }
    }

    /**
     * Handle RPC request
     * @param {Object} requestEnvelope 
     */
    async _handleRequest(requestEnvelope) {
        let responsePayload = null;
        let error = null;

        try {
            const handler = this._methodHandlers[requestEnvelope.method];
            if (handler) {
                responsePayload = await handler(requestEnvelope.payload);
            } else {
                error = `Unknown method: ${requestEnvelope.method}`;
            }
        } catch (err) {
            error = err.message || 'Internal server error';
        }

        // Send response
        if (responsePayload && !error) {
            // Check if chunking is needed
            const effectivePayloadSize = WebViewRpcConfiguration.getEffectivePayloadSize();
            if (WebViewRpcConfiguration.enableChunking && 
                responsePayload.length > effectivePayloadSize) {
                // Send as chunks
                await this._sendChunkedMessage(
                    requestEnvelope.requestId, 
                    requestEnvelope.method, 
                    responsePayload, 
                    false, 
                    error
                );
            } else {
                // Send as single message
                const responseEnvelope = {
                    requestId: requestEnvelope.requestId,
                    isRequest: false,
                    method: requestEnvelope.method,
                    payload: responsePayload
                };
                
                const responseBytes = encodeRpcEnvelope(responseEnvelope);
                const responseBase64 = uint8ArrayToBase64(responseBytes);
                this._bridge.sendMessage(responseBase64);
            }
        } else {
            // Error response
            const responseEnvelope = {
                requestId: requestEnvelope.requestId,
                isRequest: false,
                method: requestEnvelope.method,
                error: error || 'Unknown error'
            };
            
            const responseBytes = encodeRpcEnvelope(responseEnvelope);
            const responseBase64 = uint8ArrayToBase64(responseBytes);
            this._bridge.sendMessage(responseBase64);
        }
    }

    /**
     * Send a message in chunks
     * @private
     */
    async _sendChunkedMessage(requestId, method, data, isRequest, error = null) {
        const chunkSetId = `${requestId}_${crypto.randomUUID()}`;
        const effectivePayloadSize = WebViewRpcConfiguration.getEffectivePayloadSize();
        const totalChunks = Math.ceil(data.length / effectivePayloadSize);

        for (let i = 1; i <= totalChunks; i++) {
            const offset = (i - 1) * effectivePayloadSize;
            const length = Math.min(effectivePayloadSize, data.length - offset);
            const chunkData = data.slice(offset, offset + length);

            const envelope = {
                requestId,
                isRequest,
                method,
                payload: chunkData,
                chunkInfo: {
                    chunkSetId,
                    chunkIndex: i,
                    totalChunks,
                    originalSize: data.length
                }
            };

            // Only set error on the first chunk
            if (i === 1 && error) {
                envelope.error = error;
            }

            const bytes = encodeRpcEnvelope(envelope);
            const base64 = uint8ArrayToBase64(bytes);
            this._bridge.sendMessage(base64);

            // Optional: Add small delay between chunks
            if (i < totalChunks) {
                await new Promise(resolve => setTimeout(resolve, 1));
            }
        }
    }

    /**
     * Legacy property for compatibility
     * @deprecated Use addService() instead
     */
    get services() {
        return this._services;
    }
}
